design/upgrade.md

```diff
@@ -0,0 +1,179 @@
+
+#### **Rook System Namespace**
+The Rook system namespace contains the single control plane for all Rook clusters in the environment.
+This system namespace should be upgraded first before any individual clusters are upgraded.
+
+**Operator:** The operator pod itself is upgraded first since it is the host of the upgrade controller.
+If there is any new upgrade logic or any migration needed, the new version of the upgrade controller would know how to perform it, so it needs to be updated first.
+This will be a manual operation by the admin, ensuring that they are ready for their cluster to begin the upgrade process:
+```bash
+kubectl set image deployment/rook-operator rook-operator=rook/rook:v0.6.1
+```
+This command will update the image field of the operator's pod template, which will then begin the process of the deployment that manages the operator pod to terminate the pod and start a new one running the new version in its place.
+
+**Agents:** The Rook agents will also be running in the Rook system namespace since they perform operations for all Rook clusters in the environment.
+When the operator pod comes up on a newer version than the agents, it will use the Kubernetes API to update the image field of the agent's pod template.
+After this update, it will then terminate each agent pod in a rolling fashion so that their managing daemon set will replace them with a new pod on the new version.
+
+Once the operator and all agent pods are running and healthy on the new version, the administrator is free to begin the upgrade process for each of their Rook clusters.
+
+#### **Rook Cluster(s)**
+**Each** individual Rook cluster in the environment can be upgraded independently with the following flow:
+1. The cluster administrator updates their Rook Cluster CRD to update the `versionTag` field using a command similar to `kubectl -n rook edit cluster.rook.io rook`
+1. The Rook operator receives the `update` event for the Cluster CRD instance and sees that the `versionTag` field has changed from its previous value.
+    1. If the Rook system namespace upgrade described above has not yet occurred, then the operator should reject the cluster version `update` event.  The operator should never allow a cluster's version to be newer than its own version.
+1. The upgrade controller begins a reconciliation to bring the clusters actual version value in agreement with the desired version from the cluster CRD.
+As each step in this sequence begins/ends, the status field of the cluster CRD will be updated to indicate the progress (current step) of the upgrade process.
+This will help the upgrade controller resume the upgrade if it were to be interrupted.
+Also, each step should be idempotent so that if the step has already been carried out, there will be no unintended side effects if the step is resumed or run again.
+1. **API:** The API pod will be upgraded first by updating the `image` field of the pod template spec.
```

Comment:
@travisn The API pod is no longer existent, correct?

Reply:
yes, this is an old PR. With each release of rook we have updated the upgrade documentation so we can support upgrades from the previous version of rook. I'd suggest we discuss in slack.